File: app/src/main/java/com/chess/cryptobot/market/MarketRequest.java
package com.chess.cryptobot.market;

import com.chess.cryptobot.exceptions.MarketException;
import com.chess.cryptobot.model.response.ErrorResponse;
import com.chess.cryptobot.model.response.MarketResponse;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

abstract class MarketRequest implements Market {
    final String url;
    String path;
    final String apiKey;
    String algorithm;
    private final String secretKey;
    private final String DEFAULT_ENCODING = "UTF-8";

    MarketRequest(String url, String apiKey, String secretKey) {
        this.url = url;
        this.apiKey = apiKey == null ? "" : apiKey;
        this.secretKey = secretKey == null ? "" : secretKey;
    }

    abstract Gson initGson();

    Retrofit initRetrofit(Gson gson) {
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(1, TimeUnit.MINUTES)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(15, TimeUnit.SECONDS)
                .build();

        return new Retrofit.Builder()
                .baseUrl(this.url)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
    }

    abstract Object initService(Retrofit retrofit);

    public boolean keysIsEmpty() {
        return this.apiKey.isEmpty() || this.secretKey.isEmpty();
    }

    private String encode(String value) {
        return asString(encodeToBytes(value));
    }

    private byte[] encodeToBytes(String value) {
        try {
            SecretKeySpec keySpec = new SecretKeySpec(
                    this.secretKey.getBytes(DEFAULT_ENCODING), algorithm);

            Mac mac = Mac.getInstance(algorithm);
            mac.init(keySpec);
            return mac.doFinal(value.getBytes(DEFAULT_ENCODING));

        } catch (NoSuchAlgorithmException | InvalidKeyException | UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }
    }

    private String asString(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString().toUpperCase();
    }

    private String buildQueryString(Map<String, String> args) {
        StringBuilder result = new StringBuilder();
        for (String param : args.keySet()) {
            if (result.length() > 0) result.append('&');
            try {
                result.append(URLEncoder.encode(param, DEFAULT_ENCODING))
                        .append("=").append(URLEncoder.encode(args.get(param), DEFAULT_ENCODING));
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
        return result.toString();
    }

    String makeHash(Map<String, String> queryParams) {
        return encode(String.format(Locale.US, "%s%s", this.path, buildQueryString(queryParams)));
    }

    MarketResponse execute(Call<? extends MarketResponse> call) throws MarketException {
        MarketResponse response;
        try {
            Response<?> result = call.execute();
            response = (MarketResponse) result.body();
            if (response == null) {
                ResponseBody errorBody = result.errorBody();
                if (errorBody == null) {
                    throw new MarketException("No response");
                } else {
                    Gson gson = new GsonBuilder().create();
                    ErrorResponse errorResponse = gson.fromJson(errorBody.string(), ErrorResponse.class);
                    throw new MarketException(errorResponse.getErrorMessage());
                }
            } else if (!response.success()) {
                throw new MarketException(response.message());
            }
        } catch (IOException e) {
            throw new MarketException(e.getMessage());
        }
        return response;
    }
}

File: app/src/main/java/com/chess/cryptobot/market/MarketRequest.kt
package com.chess.cryptobot.market

import com.chess.cryptobot.exceptions.MarketException
import com.chess.cryptobot.model.response.ErrorResponse
import com.chess.cryptobot.model.response.MarketResponse
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import java.util.*
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

private const val DEFAULT_ENCODING = "UTF-8"

abstract class MarketRequest(val url: String, apiKey: String?, secretKey: String?) : Market {
    var path: String? = null
    var algorithm: String? = null
    val apiKey: String = apiKey ?: ""
    private val secretKey: String = secretKey ?: ""

    abstract fun initGson(): Gson

    open fun initHttpClient(): OkHttpClient {
        return OkHttpClient.Builder()
                .connectTimeout(2, TimeUnit.MINUTES)
                .readTimeout(45, TimeUnit.SECONDS)
                .writeTimeout(20, TimeUnit.SECONDS)
                .build()
    }

    fun initRetrofit(gson: Gson): Retrofit {
        return Retrofit.Builder()
                .baseUrl(url)
                .client(initHttpClient())
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
    }



    abstract fun initService(retrofit: Retrofit): Any
    override fun keysIsEmpty(): Boolean {
        return apiKey.isEmpty() || secretKey.isEmpty()
    }

    private fun encode(value: String): String {
        return asString(encodeToBytes(value))
    }

    private fun encodeToBytes(value: String): ByteArray {
        return try {
            val keySpec = SecretKeySpec(
                    secretKey.toByteArray(charset(DEFAULT_ENCODING)), algorithm)
            val mac = Mac.getInstance(algorithm)
            mac.init(keySpec)
            mac.doFinal(value.toByteArray(charset(DEFAULT_ENCODING)))
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException(e)
        } catch (e: InvalidKeyException) {
            throw RuntimeException(e)
        } catch (e: UnsupportedEncodingException) {
            throw RuntimeException(e)
        }
    }

    private fun asString(bytes: ByteArray): String {
        val sb = StringBuilder()
        for (b in bytes) {
            sb.append(String.format("%02x", b))
        }
        return sb.toString().toUpperCase(Locale.getDefault())
    }

    private fun buildQueryString(args: Map<String, String>): String {
        val result = StringBuilder()
        for (param in args.keys) {
            if (result.isNotEmpty()) result.append('&')
            try {
                result.append(URLEncoder.encode(param, DEFAULT_ENCODING))
                        .append("=").append(URLEncoder.encode(args[param], DEFAULT_ENCODING))
            } catch (ex: Exception) {
                ex.printStackTrace()
                result.append(param).append("=").append(param)
            }
        }
        return result.toString()
    }

    fun makeHash(queryParams: Map<String, String>): String {
        return encode(String.format(Locale.US, "%s%s", path, buildQueryString(queryParams)))
    }

    @Throws(MarketException::class)
    fun execute(call: Call<out MarketResponse?>): MarketResponse {
        val response: MarketResponse?
        try {
            val result: Response<*> = call.execute()
            response = result.body() as MarketResponse?
            if (response == null) {
                val errorBody = result.errorBody()
                if (errorBody == null) {
                    throw MarketException("No response")
                } else {
                    val gson = GsonBuilder().create()
                    val errorResponse = gson.fromJson(errorBody.string(), ErrorResponse::class.java)
                    throw MarketException(errorResponse.errorMessage)
                }
            } else if (!response.success()) {
                throw MarketException(response.message())
            }
        } catch (e: Exception) {
            throw MarketException(e.message)
        }
        return response
    }

}